### PR TITLE
texlive: create scope

### PR DESCRIPTION
**Pull request by xworld21**
## Description of changes
Third step along the progression multi-output / buildEnv / scope / ???: create a `texlive` scope that can be overridden in a monolithic way via `overrideScope`. Contains the minimum bits and pieces of #250626 to get here.

This should prove that the multi-output style and `texlive.__buildEnv` are complete enough to move forward.

Note that this is only groundwork for proper overridability: we need further bits, like an overridable `buildTeXLivePackage`, and a user-friendly way to override everything related to tlpdb; #250626 has more stuff about overriding the binaries in a nice way, allowing for different years to coexist, for instance.

## Things done
- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [X] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/typesetting/tex/texlive/default.nix [hunk 1]
```diff
@@ -478,12 +478,14 @@ let
     texliveBinaries = bin;
   };
 
-  tl = lib.mapAttrs (pname: { revision, extraRevision ? "", ... }@args:
+  texlivePackages = lib.makeScope newScope (self: lib.mapAttrs (pname: { revision, extraRevision ? "", ... }@args:
```
Comment (xworld21) [inline pkgs/tools/typesetting/tex/texlive/default.nix:481]: To understand: why do we need `newScope` here (as opposed to `(x: x)`)? The only difference seems to be the attribute `callPackage`, which we do not use, after all. Everybody else seems to be using `makeScope newScope`.

**Comment by xworld21** [comment]
@apfelkuchen6 I believe I jumped ahead of you!

**Comment by apfelkuchen6** [comment]
The main benefit of the the approach in #120578  -- making the whole texlive attrset a scope -- is that it allows overriding the binaries, which isn't really possible with this approach, where just the package set is a scope.

Having overridable binaries is kind of neccessary for adding texlive-2023 without ugly case distinctions, so we will probably have to make texlive a scope anyway (I'm open to other ideas though). I'm not sure whether also making the package set a scope adds anything in this case. One could already override the package set with something along the lines of  `texlive.overrideScope (self: super: { pkgs = super.pkgs // { hyperref = super.pkgs.hyperref.override { ... }; }; }).combined.scheme-full` while maintaining self-references correctly.  When pkgs is also a scope nested in the other scope, one would have to use `texlive.overrideScope (self: super { pkgs = super.pkgs.overrideScope' (pself: psuper: { hyperref = psuper.hyperref.override { ...}; }); }).combined.scheme-full` or `texlive.combine { inherit (texlive.pkgs.overrideScope' (pself: psuper: { hyperref = psuper.hyperref.override { ...}; });) scheme-full; }`.

File: pkgs/tools/typesetting/tex/texlive/default.nix (hunk 1 above)
Comment (apfelkuchen6) [reply pkgs/tools/typesetting/tex/texlive/default.nix:481]: if we don't need/want callPackage, we could use 
```suggestion
  texlivePackages = lib.makeExtensible (self: lib.mapAttrs (pname: { revision, extraRevision ? "", ... }@args:
```
In this case, the "extender"-attribute is called `extend` instead of `overrideScope'`.

**Comment by xworld21** [comment]
> that it allows overriding the binaries, which isn't really possible with this approach, where just the package set is a scope.

I see. I suppose the key point here is that unlike lua/python/perl/etc, the binary part of texlive is not quite a monolithic package; where one would manipulate `python3` directly at the top level, here we need to reach inside `texlive`. Even worse, some binaries have sources that come from the packages themselves! Like `biber` for example. Clearly if we override the package, the binary must change in sync, so all of `texlive` really must be a single scope.

TL;DR we have to move `makeScope` up one level so that the scope includes `bin`. The remaining challenge, aside from splicing, is that `bin` is nested, but it should be easy to rewrite it as a function of `self` (in practice, of `self.bin`), so that the following
```nix
texlive.overrideScope (final: prev: { bin = prev.bin // { core = prev.bin.core.overrideAttrs { ...}; }; })
```
achieves the intended result.

Does the above make sense? (Code tomorrow probably, if you don't beat me to it.)

File: pkgs/tools/typesetting/tex/texlive/default.nix (hunk 1 above)
Comment (xworld21) [reply pkgs/tools/typesetting/tex/texlive/default.nix:481]: > "broken" callPackage and newScope attributes.

My problem is probably that I don't quite understand what `newScope` is supposed to do. Is it to make it so the new `callPackage` knows all the attributes from the previous scope?

> Edit: there also is `lib.makeExtensibleWithCustomName`.

Excellent. We want camel case attributes whenever possible to avoid a repeat of the `combine` clash, and `makeScope` injects `packages`, low risk name for sure, but still a risk. If we don't have a use for `newScope` and `callPackage`, it seems like an obvious choice.

**Comment by apfelkuchen6** [comment]
That makes sense.

Unfortunately, getting the splicing to work in combination with the compatibility fixups that reinsert the packages into the toplevel is not particular easy. Just doing what everyone else does (defining the other splices "from the top" as `pkgsBuildBuild.texlive` etc (for example with generateSplicesForMkScope) results in infinite recursion. The idea I had to work around this is to explicitly define the other splices via `pkgsBuildBuild.callPackage ./. {}` and moving the compatibility fixups to another file -- with the drawback being that modifications to texlive via overlays don't propagate to the other splices.

Can you think of a better solution?

I have code for this lying around, I hope it is somewhat rebasable :D

**Comment by xworld21** [comment]
> I have code for this lying around, I hope it is somewhat rebasable :D

Yes please!

Any suggestions for debugging the splicing? E.g. how/where can I inspect `__spliced`?

**Comment by apfelkuchen6** [comment]
> Any suggestions for debugging the splicing? E.g. how/where can I inspect `__spliced`?

You can browse in the repl. Be aware that only cross package sets are spliced:
```
❯ nix repl --file .                                                
Welcome to Nix 2.17.0. Type :? for help.

Loading installable ''...
Added 20065 variables.
nix-repl> pkgsCross.aarch64-multiplatform.__splicedPackages.texlive.bin.core.__spliced
{ buildBuild = «derivation /nix/store/i28q1gbv0nn98qqq9zz1j13aqgdgwc3g-texlive-bin-2022.drv»; buildHost = «derivation /nix/store/i28q1gbv0nn98qqq9zz1j13aqgdgwc3g-texlive-bin-2022.drv»; buildTarget = «derivation /nix/store/i28q1gbv0nn98qqq9zz1j13aqgdgwc3g-texlive-bin-2022.drv»; hostHost = «derivation /nix/store/fr90pyv3nw1ngv3yaaycwa3bwdfnrg8c-texlive-bin-aarch64-unknown-linux-gnu-2022.drv»; hostTarget = «derivation /nix/store/fr90pyv3nw1ngv3yaaycwa3bwdfnrg8c-texlive-bin-aarch64-unknown-linux-gnu-2022.drv»; }

nix-repl> 
```

No idea how to debug infinite recursion though. I just noticed that currently cross-compilation of texlive.bin.core's dependency libavif is broken, so we cannot really test it anyway, but what I've done at least instantiates.

**Comment by xworld21** [comment]
> nix-repl> pkgsCross.aarch64-multiplatform.__splicedPackages.texlive.bin.core.__spliced

Oh so it's the packages under `__splicedPackages` that get the splices (well, most of them, not `texlive.bin.core` right now)... Thanks!

**Comment by apfelkuchen6** [comment]
When accessing things via "fully qualified attr paths" such as `pkgsCross.aarch64-multiplatform.__splicedPackages.texlive.bin.core.__spliced`, everything should be spliced without any individual support by the specific derivations / scopes / whatever as it's handled from the top by `splice.nix`, which creates `__slpicedPackages` by recursing into all the attrs and adding the `__spliced` attributes to every derivation.

Adding custom splicing-magic is only neccessary when things in scope refer to each other "relatively" via the scope and not "fully qualified" via the top level. In this case, the splicing done in `__splicedPackages` is not applied yet, so the scope has to add the `__spliced` attributes for relative references itself.

But this means that there probably is no pretty way to debug whether the splicing is done correctly inside the scope, as we only have the "outside view" where `splice.nix` messes with the `__spliced` attributes (or does it? I don't fully understand all this splicing voodoo). So the only way to "correctly" debug the splicing in the scope is to check with which arguments the individual functions are actually called by inserting `builtins.trace` all over the place.

**Comment by xworld21** [comment]
I think I have figured it out? I can create a scope using builtins tools only via `makeScopeWithSplicing'` and `generateSplicesForMkScope "texlive"`. It seems to work as we want, because `generateSplicesForMkScope` recurses into attribute sets, so it reaches into `bin.core` as well. Then one needs to sprinkle `self` all over the place, and trust that `overrideScope` is doing its intended job, i.e. it applies the overrides to the splices.

*Except* that`generateSplicesForMkScope` does *not* traverse lists, so our dear `{ pkgs = [ ... ]; }` gets in the way. *Again*. But if we expose the packages as multi-output-like sets under `texlive.pkgs`, and make `{ pkgs = [ ... ]; }` a compatibility shim, we should be fine in theory?

Give me a moment to complete the multi-output, then I'll rebase this PR, and we'll compare against your reference implementation.

File: pkgs/tools/typesetting/tex/texlive/default.nix [hunk 2]
```diff
@@ -564,4 +570,18 @@ in
...
+  };
+in
+# backward compatibility layer: expose packages directly under texlive, but not as derivations
+# this cannot be done from within the scope
+(builtins.mapAttrs
+  # flag if manpages are present
+  (n: p: let pWithMan = p // lib.optionalAttrs (p ? man) { texdoc = p.texdoc // { hasManpages = true; }; }; in {
+    pkgs = builtins.concatMap (n: lib.optional (p ? ${n}) p.${n})  [ "tex" "texdoc" "texsource" "tlpkg" "out" ];
+  })
+  texlive.__pkgs) // texlive
```
Comment (xworld21) [inline pkgs/tools/typesetting/tex/texlive/default.nix:587]: The compromise here is that you cannot use `inherit (texlive.overrideScope ...) package`, so this must come after `buildEnv`.

Comment (xworld21) [reply pkgs/tools/typesetting/tex/texlive/default.nix]: ...or we do @apfelkuchen6's `applyOverScope` trick:
https://github.com/NixOS/nixpkgs/pull/250626/commits/13e78086cee9dbfb1d9dcb531cbbaf899987c52c#diff-78dd1e8609118bb30c483b837f017110017218a9b7c388887a9c0ed0aa823f36R162-R176

File: pkgs/tools/typesetting/tex/texlive/bin.nix [hunk 3]
```diff
@@ -90,11 +90,12 @@ core = stdenv.mkDerivation rec {
 
   nativeBuildInputs = [
     pkg-config
-  ] ++ lib.optionals (!stdenv.buildPlatform.canExecute stdenv.hostPlatform) [
+  ] ++ lib.optionals (!stdenv.buildPlatform.canExecute stdenv.hostPlatform && self.bin.core ? __spliced) [
```
Comment (xworld21) [inline pkgs/tools/typesetting/tex/texlive/bin.nix:93]: @apfelkuchen6 this was causing infinite recursion when evaluating the splice `hostHost`, because in that case the build `self.bin.core` would be the same as the host `self.bin.core`, regardless of the true `buildPlatform` and `hostPlaftorm`. Conveniently, `__spliced` is not generated in `hostHost`, hence the `? __spliced` guard.

I don't think I have seen this pattern anywhere. Is there a more idiomatic way to address the issue?

(This is an academic issue as we would never evaluate `hostHost` in the actual build, it was just annoying for debugging.)

Comment (apfelkuchen6) [reply pkgs/tools/typesetting/tex/texlive/bin.nix:93]: There are a bunch of derivations with cross self-dependenies (for example vala and guile), but I couldn't find anything that only adds itself if the derivation is spliced.

I'd consider it a non-issue. Maybe doing nothing about it is even better as it might detect actual problems at eval time.
